src/wrappers.jl

```diff
@@ -1,32 +1,90 @@
 
-#TODO: composed preconditioners, preconditioner setter for cache,
-#   detailed tests for wrappers
-
 ## Preconditioners
 
+"""
+ P  *  x = x .* (1/P.s)
+
+ Pi *  x = x .* (P.s)
+"""
 struct ScaleVector{T}
     s::T
     isleft::Bool
 end
 
-function LinearAlgebra.ldiv!(v::ScaleVector, x)
+# y = A x
+function LinearAlgebra.mul!(y, A::ScaleVector, x)
+    A.s == one(eltype(A.s)) && return y = x
+
+    if A.isleft
+        @. x = x / A.s
+    else
+        @. x = x * A.s
+    end
+end
+
+# A B α + C β
+function LinearAlgebra.mul!(C, A::ScaleVector, B, α, β)
+    
+    tmp = zero(B)
+    C = β * C + α * mul!(tmp, A, B)
```

Comment:
This is both allocating and shadows C so it's incorrect. Split the operations to first use the tmp and then mutate C.

Reply:
would this be better?

```julia
# A B α + C β
function LinearAlgebra.mul!(C, A::ScaleVector, B, α, β)
    A.s == one(eltype(A.s)) && return @. C = α * B + β * C

    if A.isleft
        @. C = β * C + (α / A.s) * B
    else
        @. C = β * C + (α * A.s) * B
    end
end
```

Reply:
i think this would be faster:
```julia
# A B α + C β
function LinearAlgebra.mul!(C, A::ScaleVector, B, α, β)
    A.s == one(eltype(A.s)) && return @. C = α * B + β * C

    s = A.isleft ? 1/A.s : A.s
    mul!(C, s, B, α, β)
end 
```